scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala

```diff
@@ -0,0 +1,153 @@
+            val forwardCompatRepeated = ParquetListFormatRule.wrapElementAsRepeatedType(
+              sourceRule,
+              sourceRepeatedListType,
+              forwardCompatElement
+            )
+            ParquetListFormatRule.formatForwardCompatibleRepeatedType(
+              forwardCompatRepeated,
+              targetRepeatedListType)
+          }
+          case _ => sourceRepeatedListType // No-op
+        }
+        // Wrapped the formatted repeated type in its original groups,
+        // describing field name and whether it's optional/required
+        sourceWrappers.foldRight(formattedRepeated) {
+          (wrapper, group) => wrapper.withNewFields(group)
+        }
+      case (
+        GroupUnwrapped(sourceWrappers, None, Some(sourceRepeatedMapType)),
+        GroupUnwrapped(_, None, Some(targetRepeatedMapType))
+      ) =>
+        // Format map
+        val forwardCompatRepeated = formatForwardCompatibleType(sourceRepeatedMapType, targetRepeatedMapType)
+        val formattedRepeated = ParquetMapFormatRule.formatForwardCompatibleRepeatedType(
+          forwardCompatRepeated,
+          targetRepeatedMapType
+        )
+        // Wrapped the formatted repeated type in its original groups,
+        // describing field name and whether it's optional/required
+        sourceWrappers.foldRight(formattedRepeated) {
+          (wrapper, group) => wrapper.withNewFields(group)
+        }
+      case _ =>
+        // Field projection
+        val sourceGroup = sourceType.asGroupType
+        val targetGroup = targetType.asGroupType
+
+        import scala.collection.JavaConverters._
+        val resultFields = sourceGroup.getFields.asScala.map { sourceField =>
+          if (!targetGroup.containsField(sourceField.getName)) {
+            if (!sourceField.isRepetition(Repetition.OPTIONAL)) {
+              throw new DecodingSchemaMismatchException(
+                s"Found non-optional source field ${sourceField.getName}:\n$sourceField\n\n" +
+                  s"not present in the given target type:\n${targetGroup}"
+              )
+            }
+            sourceField
+          }
+          else {
+            val fieldIndex = targetGroup.getFieldIndex(sourceField.getName)
+            val targetField = targetGroup.getFields.get(fieldIndex)
+            formatForwardCompatibleType(sourceField, targetField)
+          }
+        }
+        sourceGroup.withNewFields(resultFields.asJava)
+    }
+  }
+
+  private case class GroupUnwrapped(wrappers: Seq[GroupType],
+                                    repeatedListType: Option[Type] = None,
+                                    repeatedMapType: Option[Type] = None)
+
+  private def unwrapGroup(typ: Type, wrappers: Seq[GroupType] = Seq()): GroupUnwrapped = {
+    if (typ.isPrimitive) {
+      GroupUnwrapped(
+        wrappers,
+        repeatedListType=None,
+        repeatedMapType=None
+      )
+    } else if (typ.asGroupType.getFieldCount != 1) {
```

Comment:
In this `else if` branch and the `else` branch I noticed we have `typ.asGroupType` calls everywhere. To avoid that and to make it clear that `!typ.isPrimitive` implies that `typ` is a group type, could we instead do something like

```scala
if (typ.isPrimitive) {
   // ...
} else {
  val groupTyp = typ.asGroupTyp
  if (groupTyp.getFieldCount != 1) {
     //...
  } else {
     //...
  }
}
```

We might be able to further improve clarity by changing the sign of the second condition, doing

```scala
if (typ.isPrimitive) {
   // ...
} else {
  val groupTyp = typ.asGroupTyp
  if (groupTyp.getFieldCount == 1) {  // <--- inverted condition
     //...
  } else {
     //...
  }
}
```

which then simplifies or clarifies the

```
// Note the field count is strictly 1 here
```

comment present in the current code.

We could then add a comment in the `else` branch saying something like

```
// The group contains more than one field, so ... <whatever that implies>
```

This is maybe nitpicky, but I'm trying to walk through each case here to make sure that I understand what all of the branches correspond to, so I'm considering whether a different ordering or presentation of the cases can make it easier to understand and verify.

Reply:
Simplified this to two methods to extract group list and maplist instead.